test/fgpu/test_engine.py

```diff
@@ -933,6 +937,60 @@ async def test_missing_heaps(
             == np.sum(~dst_present) * n_substreams
         )
 
+        # Information needed to check that the dig-rms-dbfs sensor is correctly
+        # going into FAILURE when there are missing samples.
+        time_converter = TimeConverter(SYNC_TIME, ADC_SAMPLE_RATE)
+        unix_timestamps = [time_converter.adc_to_unix(t) for t in timestamps]
+
+        # The sensor is updated once per out_item.
+        spectra_per_output_chunk = engine_server.chunk_jones // output.channels
+        heaps_per_output_chunk = spectra_per_output_chunk // spectra_per_heap
+
+        # Need to reshape timestamp list so that we can work in batches of
+        # chunk size. The last chunk isn't finished so we need to add a
+        # bit on the end.  Need to test this after the prometheus counters
+        # becaus this affects np arrays which those tests use to count things.
+        if (remainder := len(unix_timestamps) % spectra_per_output_chunk) != 0:
+            unix_timestamps.extend([0 for _ in range(spectra_per_output_chunk - remainder)])
+        chunk_timestamps = np.reshape(unix_timestamps, (-1, spectra_per_output_chunk))
+        chunk_time_step = np.average(chunk_timestamps[1]) - np.average(chunk_timestamps[0])
+
+        # Same with dst_present
+        if (remainder := len(dst_present) % heaps_per_output_chunk) != 0:
+            dst_present = np.append(dst_present, [False for _ in range(heaps_per_output_chunk - remainder)])
+        chunk_present = np.reshape(dst_present, (-1, heaps_per_output_chunk))
+
+        # Sensor is not present in the narrowband mode.
+        if output.decimation == 1:
+            # We can't just for-loop because we may not have an update every chunk.
+            sensor_update_iter = iter(sensor_update_dict["input1.dig-rms-dbfs"])  # Both sensors should be the same.
+            sensor_update = next(sensor_update_iter)
+
+            for chunk_time, present in zip(chunk_timestamps, chunk_present):
+                # Sensor timestamp should be close to the end of the chunk if # it exists.
+                diff = abs(sensor_update.timestamp - np.max(chunk_time))
```

Comment:
Using `abs` concerns me - if the sensor timestamp is in the past then this condition will continue to be true and we'll never advance to the next sensor reading.